trusted/crypto/src/lib.rs

```diff
@@ -37,6 +56,78 @@ impl Sha256 {
     }
 }
 
+/// Builds a [Sha256] hash in a piecewise fashion.  This can be used when all
+/// of the data to hash is not available in contiguous memory.
+pub struct Sha256Builder {
+    handle: sgx_sha_state_handle_t,
+}
+
+impl Sha256Builder {
+    /// Create a new builder
+    ///
+    /// # Returns
+    /// An empty [Sha256Builder] ready to compute a hash.
+    /// See [Sha256Builder::update()].
+    ///
+    /// # Errors
+    ///
+    /// If this function will return an error when:
+    /// - there is not enough memory in SGX
+    /// - there is an internal library error in SGX
+    pub fn new() -> Result<Self, Error> {
+        let mut handle: sgx_sha_state_handle_t = ptr::null_mut();
+        let result = unsafe { sgx_sha256_init(&mut handle) };
+        match result {
+            sgx_status_t::SGX_SUCCESS => Ok(Sha256Builder { handle }),
+            x => Err(Error::SGX(x)),
+        }
+    }
+
+    /// Update the data hashed so far
+    ///
+    /// # Arguments
+    /// - `data` The data to use in updating the hash.
+    ///
+    /// # Returns
+    /// The updated [Sha256Builder].
+    ///
+    /// # Errors
+    ///
+    /// This function will return an error if there is an internal library error
+    /// in SGX.
+    pub fn update(self, data: &[u8]) -> Result<Self, Error> {
+        let result =
+            unsafe { sgx_sha256_update(data.as_ptr(), data.len().try_into()?, self.handle) };
+        match result {
+            sgx_status_t::SGX_SUCCESS => Ok(self),
+            x => Err(Error::SGX(x)),
+        }
+    }
+
+    /// Get the final hash of all the data seen in [Sha256Builder::update()]
+    ///
+    /// # Returns
+    /// The [Sha256] hash of all the `data`, in order, provided in
+    /// [Sha256Builder::update()]
+    ///
+    /// # Errors
+    ///
+    /// This function will return an error if there is an internal library error
+    /// in SGX.
+    pub fn build(self) -> Result<Sha256, Error> {
+        let mut hash: sgx_sha256_hash_t = Default::default();
+        let result = unsafe { sgx_sha256_get_hash(self.handle, &mut hash) };
+        // The errors from sgx_sha256_close() are success or a NULL pointer.
+        // There isn't much error handling we can do if for some reason
+        // self.handle is a NULL pointer.
+        unsafe { sgx_sha256_close(self.handle) };
```

Comment:
IIUC we should move this Close call into an `impl Drop for Sha256Builder`, or else errors from `update()` will result in unclosed handles.

Reply:
oh good catch!

Reply:
added a drop method